code/controllers/subsystems/lighting.dm

```diff
@@ -100,59 +110,88 @@ var/datum/controller/subsystem/lighting/SSlighting
 	var/list/curr_corners = corner_queue
 	var/list/curr_overlays = overlay_queue
 
-	while (curr_lights.len)
-		var/datum/light_source/L = curr_lights[curr_lights.len]
-		curr_lights.len--
+	while (curr_lights.len && lq_idex <= curr_lights.len)
+		var/datum/light_source/L = curr_lights[lq_idex++]
 
-		L.update_corners()
+		if (L.needs_update != LIGHTING_NO_UPDATE)
+			L.update_corners()
 
-		L.needs_update = LIGHTING_NO_UPDATE
+			L.needs_update = LIGHTING_NO_UPDATE
 
-		processed_lights++
+			processed_lights++
 
 		if (no_mc_tick)
 			CHECK_TICK
 		else if (MC_TICK_CHECK)
 			break
 
+	if (lq_idex > 1)
+		curr_lights.Cut(1, lq_idex)
+		lq_idex = 1
+
 	if (!no_mc_tick)
 		MC_SPLIT_TICK
 
-	while (curr_corners.len)
-		var/datum/lighting_corner/C = curr_corners[curr_corners.len]
-		curr_corners.len--
+	while (curr_corners.len && cq_idex <= curr_corners.len)
```

Comment:
Same story here as with the above while condition. Since cq_idex >= 1 always, you can opt out of checking len.